### ISSUE TITLE
JsonLogger should call Stringer when logging values too

### ISSUE THREAD

**Issue by hydroflame** [go-kit/kit#147]
json.Encode doesn't do that and people are more likely to implement fmt.Stringer then json.Marshaler. 

issues to consider:
 - should it favor json.Marshaler before fmt.Stringer

json_logger.go

    func merge(dst map[string]interface{}, k, v interface{}) {
	    var key string
    	switch x := k.(type) {
    	case string:
    		key = x
    	case fmt.Stringer:
    		key = safeString(x)
    	default:
    		key = fmt.Sprint(x)
	    }
    	if x, ok := v.(error); ok {
    		v = safeError(x)
   	}
    
	    switch x := v.(type) {
    	case string:
    		v = x
    	case fmt.Stringer:
    		v = safeString(x)
    	default:
    		v = fmt.Sprint(x)
    	}
    	if x, ok := v.(error); ok {
    		v = safeError(x)
    	}
   	dst[key] = v
    }

**Comment by ChrisHines** [comment]
I agree with @hydroflame that having JsonLogger work with `fmt.Stringer` values is probably what people expect. But @peterbourgon is right that we don't want to just `fmt.Sprint` all values before passing them down to `json.Marshal`. The code also needs to be careful about the values that might implement `json.Marshaler` or `encoding.TextMarshaler`, both of which are handled by `json.Marshal` already. I think those interfaces should take priority over `fmt.Stringer`. Also, although rare, what if a type implements `error` and `fmt.Stringer`? I suggest code like:

```go
switch x := v.(type) {
case json.Marshaler:
case encoding.TextMarshaler:
case error:
    v = safeError(x)
case fmt.Stringer:
    v = safeString(x)
}
dst[key] = v
```

**Comment by peterbourgon** [comment]
WIthout a default clause, there's no need for the first two cases, as the net effect is a no-op, right?

**Comment by ChrisHines** [comment]
The first two clauses guard against calling `safeString(x)` for a value that also implements one of those interfaces, but delegating the work of calling those interface methods to `json.Marshal`.

**Comment by peterbourgon** [comment]
Ah, I understand, now.

**Comment by ChrisHines** [comment]
I suppose the final code should include a comment explaining what's going on.

**Comment by hydroflame** [comment]
Yeah sorry I didn't meant to imply that my implementation should be used as is, I just wanted to show that it should be pretty easy to implement

**Comment by hydroflame** [comment]
I +1 @ChrisHines solution (json.Marshaller>encoding.TextMarshaller>error>fmt.Strigner)

**Comment by ChrisHines** [comment]
@hydroflame Can you add a test and put together a PR for this?

**Comment by hydroflame** [comment]
closed with #156
